### PR TITLE
Add [Discourse] status/stats shield

### PR DESCRIPTION
**Pull request by cavneb**
References https://github.com/badges/shields/issues/1325

API:

`/discourse/SCHEME/DISCOURSE_HOST/topics.svg`
`/discourse/SCHEME/DISCOURSE_HOST/posts.svg`
`/discourse/SCHEME/DISCOURSE_HOST/users.svg`
`/discourse/SCHEME/DISCOURSE_HOST/likes.svg`
`/discourse/SCHEME/DISCOURSE_HOST/status.svg`

Preview:

screen shot 2017-12-05 at 3 10 07 pm

### CODE REVIEW

File: service-tests/discourse.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  .intercept(nock => nock('https://meta.discourse.org')
+    .get('/site/statistics.json')
+    .reply(200, data)
+  )
+  .expectJSON({ name: 'discourse', value: 'invalid' });
+
+t.create('Connection Error')
+  .get('/meta.discourse.org/status.json')
+  .networkOff()
+  .expectJSON({ name: 'discourse', value: 'invalid' });
```
Comment (paulmelnikow) [inline service-tests/discourse.js]: These tests look great!

Could you also add some tests that hit the live service? We have lots of permissive regexes that accept any possible positive number, for example. The reason for this is so we can detect breaking changes that occur in the upstream API, and even run smoke tests on our live server.

File: server.js [hunk 2]
```diff
@@ -3063,6 +3063,85 @@ cache(function(data, match, sendBadge, request) {
...
+        case 'users':
+          statCount = data.user_count;
+          badgeData.text[1] = metric(statCount) + ' users';
+          break;
+        case 'likes':
+          statCount = data.like_count;
+          badgeData.text[1] = metric(statCount) + ' likes';
+          break;
+        case 'status':
+          statCount = data.like_count;
```
Comment (paulmelnikow) [inline server.js]: Remove this? ^^

File: server.js [hunk 3]
```diff
@@ -3063,6 +3063,85 @@ cache(function(data, match, sendBadge, request) {
...
+  };
+
+  var badgeData = getBadgeData('discourse', data);
+  request(options, function(err, res) {
+    if (err != null) {
+      if (res) {
+        console.error('' + res);
+      }
+
+      badgeData.text[1] = 'invalid';
```
Comment (paulmelnikow) [inline server.js]: **inaccessible** is probably more accurate here.

File: server.js [hunk 4]
```diff
@@ -3063,6 +3063,85 @@ cache(function(data, match, sendBadge, request) {
...
+      sendBadge(format, badgeData);
+      return;
+    }
+
+    if (res.statusCode !== 200) {
+      badgeData.text[1] = 'offline';
```
Comment (paulmelnikow) [inline server.js]: When does this case occur?

What happens for a nonexistent host? Is it possible to detect and return **host not found** for that case? A helpful error message makes it easier for users to understand what has gone wrong with their badge.

File: lib/all-badge-examples.js [hunk 5]
```diff
@@ -310,6 +310,46 @@ const allBadgeExamples = [
         title: 'continuousphp',
         previewUri: '/continuousphp/git-hub/doctrine/dbal/master.svg'
       },
+      {
+        title: 'Discourse topics',
+        previewUri: '/discourse/meta.discourse.org/topics.svg',
```
Comment (paulmelnikow) [inline lib/all-badge-examples.js]: Do Discourse servers always run `https`?

File: lib/all-badge-examples.js [hunk 6]
```diff
@@ -310,6 +310,46 @@ const allBadgeExamples = [
...
+        keywords: [
+          'discourse'
+        ]
+      },
+      {
+        title: 'Discourse posts',
+        previewUri: '/discourse/meta.discourse.org/posts.svg',
+        exampleUri: '/discourse/DISCOURSE_HOST/posts.svg',
+        keywords: [
+          'discourse'
```
Comment (paulmelnikow) [inline lib/all-badge-examples.js]: Since this word is in the title, you don't need to add it to keywords.

File: lib/all-badge-examples.js (hunk 5 above)
Comment (cavneb) [reply lib/all-badge-examples.js]: This is a good question. Let me find out.

File: server.js (hunk 4 above)
Comment (paulmelnikow) [reply server.js]: In retrospect, I realize the **host not found** suggestion doesn't make sense. Probably this one should also be **inaccessible**.

File: lib/all-badge-examples.js (hunk 5 above)
Comment (cavneb) [reply lib/all-badge-examples.js]: I will switch this to http. The library itself should handle the redirect if needed.

Comment (paulmelnikow) [reply lib/all-badge-examples.js]: If https is always supported, defaulting to that would be better. If https is sometimes unavailable, the way we'd handle that is to include the scheme in the badge URL.

Ideal would be: always use https, with no option for http, just like you have. Let's stick with that if it's always supported.

Comment (cavneb) [reply lib/all-badge-examples.js]: https is not guaranteed, but is typically the case. I will see how it's done currently by using the badge URL to determine the scheme.

**Comment by cavneb** [comment]
@paulmelnikow I made the requested changes. I also removed the exampleURI's. Having the previewUri is more descriptive I on usage (I think).

**Comment by paulmelnikow** [comment]
Looks good, thanks!

**Comment by cavneb** [comment]
@paulmelnikow how soon should this appear on https://shields.io?

**Pull request merged**